jenkins-master/jobs/scripts/workspace/submission.py

```diff
@@ -0,0 +1,400 @@
+#!/usr/bin/env python
+
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import argparse
+import os
+import re
+import socket
+import time
+from urlparse import urljoin, urlparse
+import uuid
+
+from config import config
+
+here = os.path.dirname(os.path.abspath(__file__))
+
+LOOKUP_FRAGMENT = 'api/project/{repository}/resultset/?revision={revision}'
+REVISON_FRAGMENT = '/#/jobs?repo={repository}&revision={revision}'
+
+BUILD_STATES = ['running', 'completed']
+
+
+class JenkinsDefaultValueAction(argparse.Action):
+    """Jenkins cannot handle optional arguments to a CLI script so we pass in
+    a 'None' value by default.
+
+    """
+    def __init__(self, option_strings, dest, nargs=None, **kwargs):
+        if nargs is not None:
+            raise ValueError("nargs not allowed")
+        argparse.Action.__init__(self, option_strings, dest, **kwargs)
+
+    def __call__(self, parser, namespace, values, option_string=None):
+        if type(values) is str:
+            values = values if values != 'None' else None
+        elif type(values) is list:
+            values = [value if value != 'None' else None for value in values]
+
+        setattr(namespace, self.dest, values)
+
+
+class FirefoxUIResultParser(object):
+
+    BUSTED = 'busted'
+    SUCCESS = 'success'
+    TESTFAILED = 'testfailed'
+    UNKNOWN = 'unknown'
+
+    def __init__(self, retval, log_file):
+        self.retval = retval
+        self.log_file = log_file
+        self.failure_re = re.compile(r'(^TEST-UNEXPECTED-FAIL|TEST-UNEXPECTED-ERROR)|'
+                                     r'(.*CRASH: )|'
+                                     r'(Crash reason: )')
+        self.failures = []
+        self.parse()
+
+    def parse(self):
+        try:
+            with open(self.log_file, 'r') as f:
+                for line in f.readlines():
+                    if self.failure_re.match(line):
+                        self.failures.append(line)
+        except IOError:
+            print('Cannot parse failures due to missing log file: {}'.format(self.log_file))
+
+    @property
+    def status(self):
+        status = self.UNKNOWN
+
+        # retval.txt was not written - so most likely an abort
+        if self.retval is None or (self.retval and not self.failures):
+            status = self.BUSTED
+
+        elif not self.failures:
+            status = self.SUCCESS
+
+        elif self.failures:
+            status = self.TESTFAILED
+
+        return status
+
+    def failures_as_json(self):
+        failures = {'all_errors': [], 'errors_truncated': True}
+
+        for failure in self.failures:
+            failures['all_errors'].append({'line': failure, 'linenumber': None})
```

Comment:
`'linenumber': None` prevent the failure summary from being displayed in the Treeherder log viewer. You can use `1` for all failures instead of `None` to get around this for now.